### PR TITLE
Issue 57: Change password page and hash passwords

### PR DESCRIPTION
**Pull request by hitbox**
- Added change password button on "Edit Profile" page.
- Password change requires re-login. (do we want that?)
- Passwords are now hashed on the Python side--meaning we can't do password comparisons in the SQLAlchemy queries (see tests).
- Require stronger passwords?
- Added reset password feature. The staff member must have an email address. The link is in the nav bar. Probably want it somewhere else.
- `migration/01_create_passwordreset.py` for bringing the server database up to support password reset. (just run `models.py` for debugging)
- Added `migration/00_hashpassword.py` to hash server  passwords instead of verifying and changing when they login. Couldn't think of a good way to do it that wasn't already done better in [SQLAlchemy-Utils](https://github.com/kvesteri/sqlalchemy-utils).

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,16 +1,17 @@
-blinker==1.4
-dominate==2.1.17
-Flask==0.10.1
 Flask-Bootstrap==3.3.5.7
```
Comment (hitbox) [inline requirements.txt:1]: Sorted this file before I realized it would come into the repo as a bunch of changes.

File: tests.py [hunk 2]
```diff
@@ -141,8 +141,14 @@ def test_adduser(self):
...
             emailaddress="test@testing.com"),follow_redirects=True)
-        staff=models.Staff.query.filter(and_(models.Staff.username=='t', models.Staff.password=='t', models.Staff.f_name=='t', models.Staff.l_name=='t', models.Staff.phonenumber==1112223333, models.Staff.emailaddress=='test@testing.com')).first()
+        staff=models.Staff.query.filter(
+            and_(models.Staff.username=='t', models.Staff.f_name=='t',
+                 models.Staff.l_name=='t',
+                 models.Staff.phonenumber==1112223333,
+                 models.Staff.emailaddress=='test@testing.com')).first()
+        #NOTE: password must be verified on the Python side now
         self.assertIsNotNone(staff)
+        self.assertTrue(staff.password=='t')
```
Comment (hitbox) [inline tests.py:157]: Password verification has to be done on the class instance now.

File: library.py [hunk 3]
```diff
@@ -31,6 +31,9 @@
 mail = Mail(app)
 bootstrap = Bootstrap(app)
 
+def logout_user():
+    session.pop('logged_in', None)
+
 def login_required(test):
```
Comment (hitbox) [inline library.py:38]: Thought a function would be easier to remember and there's two places the code is logging out now.

File: library.py [hunk 4]
```diff
@@ -443,6 +446,34 @@ def contact_status(PCID):
...
+        if user is None:
+            flash("Server error")
+            return redirect(url_for('main'))
+
+        if user.password != request.form.get('oldpass'):
+            flash("Password verification failed")
+        else:
+            newpass = request.form.get('newpass', '').strip()
+            #XXX: password requirements?
+            #     validate decorator in Staff
```
Comment (hitbox) [inline library.py]: Thinking password validation would be better in the database model.

File: models.py [hunk 5]
```diff
@@ -178,6 +180,7 @@ def init_models(session):
 def ArgParser():
     parser = argparse.ArgumentParser(description=main.__doc__)
     _a = parser.add_argument
+    _a('--noconfirm', action='store_true', help='do not ask for confirmation')
```
Comment (hitbox) [inline models.py:224]: Tired of typing y all the time.

File: passwordtype.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import bcrypt
```
Comment (hitbox) [inline passwordtype.py:1]: New requirement.

File: passwordtype.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+from sqlalchemy import types
+
+def encode(s):
+    return s.encode('utf8')
+
+def hashpassword(password):
```
Comment (hitbox) [inline passwordtype.py:7]: bcrypt was complaining and seemed to want encoding on everything.

File: passwordtype.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    return bcrypt.hashpw(encode(password), bcrypt.gensalt())
+
+class Password(object):
+
+    def __init__(self, hashed):
+        self.hashed = hashed
```
Comment (hitbox) [inline passwordtype.py:13]: Does not take the password text because `process_bin_param` processes the database values in and out; and if I did the hashing here, I was hashing hashes and passwords wouldn't verify.

File: templates/changepassword.html [hunk 9]
```diff
@@ -0,0 +1,15 @@
+{% extends "base.html" %}
```
Comment (hitbox) [inline templates/changepassword.html:1]: Simple, modified copy of `login.html`

**Comment by Stamates** [comment]
I couldn't get bcrypt to install/work.  Here's what I tried:
1. Pip install -r requirements (fail on bcrypt... "Unable to find vcvarsall.bat")
2. Pip install bcrypt (fail "Unable to find vcvarsall.bat")
3. Pip install bcrypt out of virtual environment (local).  It installed (this is usually the fix when I have the "Unable to find vcvarsall.bat" error, when I try to run python models.py in the virtual environment it fails saying it can't find bcrypt.
4. Copied bcrypt folder from local site_packages to site_packages for virtual environment.  Still failed when running models.py (looking for six).  Pip installed six, still failed now can't find _bcrypt from import statement in bcrypt__init__,py file.

You can probably chalk this up to "it sucks to run dev in Windows" but I've never had this issue where I couldn't get the package to work at all.  It would probably work if I stopped using the virtual environment and pip install -r requirements.txt to local, but I don't want to mess anything else up (which is why I like to use virtual env).

Long story short... I couldn't test this branch in order to confirm for a merge.

File: migration/01_create_passwordreset.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+except ImportError:
+    app, db, PasswordReset = None, None, None
+
+def create_passwordreset():
+    """
+    Create the passwordreset table in the database.
+    """
+    with app.app_context():
+        conn = db.session.connection()
+        table = PasswordReset.metadata.tables['passwordreset']
```
Comment (hitbox) [inline migration/01_create_passwordreset.py:17]: Just found I could have used `PasswordReset.__tablename__` instead of `'passwordreset'`.
